### DISCUSSION TITLE
Raspberry Pi Pico port

### DISCUSSION THREAD

**Discussion by ginolig** [jrowberg/i2cdevlib#615] [Show and tell]
Hello there!
I've ported the I2Cdev library as well as the MPU6050 one (including the DMP v6.12) to the [RP2040 dev board](https://www.raspberrypi.org/products/raspberry-pi-pico/) and would like to share it but I don't know how ...

**Comment by eadf** [comment]
A guide I found by googling "how to contribute on GitHub": https://www.dataschool.io/how-to-contribute-on-github/

Having a good gui for git also makes the process easier: https://www.sourcetreeapp.com https://desktop.github.com

I don't know how integrated your solution is with the "Arduino" platform, maybe the best way is to create a new root platform.

**Comment by jrowberg** [comment]
That's great! The guide @eadf posted has all the details you need, but here's the short version for how to get it integrated into the existing codebase:

1. Fork the `i2cdevlib` repo using the "Fork" button near the top right of the main `i2cdevlib` repo page
2. Clone your forked repo wherever you want on your computer
3. Add the `RP2040` folder to the repository root, and the `I2Cdev` and `MPU6050` subfolders inside that with your code
4. Commit your changes (ideally separated into multiple parts like core and device library commits, but that's optional)
5. Push your changes up to your forked repo on Github
6. On Github, use the "Contribute" link on your own main repository page to submit a pull request back to the original repo

Then we can manage your PR from this end and merge it, assuming no issues.

**Reply by jrowberg** [reply]
...to clarify, creating the "RP2040" folder in the root assumes that this is more of a low-level port, and not something meant to operate inside the Arduino IDE (assuming that's even possible).

**Reply by eadf** [reply]
I would recommend a step 2.5: create a local branch, and add your changes to that.
This will free up your local master branch for updates. 

@jrowberg May I suggest a pinned Q&A post with your guide?

**Reply by ginolig** [reply]
> ...to clarify, creating the "RP2040" folder in the root assumes that this is more of a low-level port, and not something meant to operate inside the Arduino IDE (assuming that's even possible).

Basically I've replaced all the Arduino I2C related functions with the ones from the Pico C/C++ SDK, and did some modifications needed for it to work properly (e.g. change delay() for sleep_ms() and things like that). Supposedly Arduino IDE support for the RP2040 is on its way but I'm using VS Code and cmake, as the official guides recommend. 
I wonder if that's okay for being considered a port?

**Reply by jrowberg** [reply]
Yes, that's perfectly fine for a port. If it were Arduino-specific, then a port might not be appropriate since it would be better to modify the existing Arduino code in whatever way is needed to make it work with the RP2040 board in the Arduino IDE. But since you've done it to work with the RP2040 toolchain outside of Arduino, then a port is ideal.

**Reply by ginolig** [reply]
I've created a new branch, made changes there and a pull request. Comments are more than welcome. By the way, I haven't touched the comments header on the files that had one, I'm not sure if I should add anything there...
